client/my-sites/themes/themes-selection.jsx

```diff
@@ -218,20 +217,64 @@ class ThemesSelection extends Component {
 		return options;
 	};
 
+	isMatchingTheme = ( theme ) => {
+		const { query } = this.props;
+		return (
+			theme.name?.toLowerCase() === query.search?.toLowerCase() ||
+			theme.id?.toLowerCase() === query.search?.toLowerCase()
+		);
+	};
+
+	getThemes = () => {
+		const { themes, wpOrgThemes, query, isLastPage } = this.props;
+
+		const themeSlugs = themes.map( ( theme ) => theme.id );
+		const validWpOrgThemes = wpOrgThemes.filter(
+			( wpOrgTheme ) =>
+				! themeSlugs.includes( wpOrgTheme?.id?.toLowerCase() ) && // Avoid duplicate themes. Some free themes are available in both wpcom and wporg.
+				! RETIRED_THEME_SLUGS_SET.has( wpOrgTheme?.id?.toLowerCase() ) // Avoid retired themes.
+		);
+
+		const matchingTheme = themes.find( this.isMatchingTheme );
+		const restThemes = matchingTheme
+			? themes.filter( ( theme ) => theme.id !== matchingTheme.id )
+			: themes;
+		const matchingWpOrgTheme = validWpOrgThemes.find( this.isMatchingTheme );
+		const restWpOrgThemes = matchingWpOrgTheme
+			? validWpOrgThemes.filter( ( theme ) => theme.id !== matchingWpOrgTheme.id )
+			: validWpOrgThemes;
+
+		return [
+			...( matchingTheme ? [ matchingTheme ] : [] ),
+			...( matchingWpOrgTheme ? [ matchingWpOrgTheme ] : [] ),
+			...restThemes,
+			// Only include WP.org themes when searching a term and after all previous themes are visible.
+			...( query.search && isLastPage ? restWpOrgThemes : [] ),
+		];
+	};
```

Comment:
The theme selection shouldn't be responsible for implementing the mix and merge strategy since its primary responsibility is orchestrating the necessary components to show the theme selection screen (correct me if I'm wrong on this last part, but regardless), maybe we could move the logic to a repository so that we don't couple the strategy for mixing and merging themes to the theme-selection screen. 

We could then also add some nice unit tests and increase our coverage of core theme logic.

Reply:
Thanks for you feedback! Done in https://github.com/Automattic/wp-calypso/pull/78174/commits/09b6ad12c03030facdaabb763838cfa3a4ce1ac8 (and tests added in https://github.com/Automattic/wp-calypso/pull/78174/commits/1627d78e5f0e24d89959a663df3757f891d70d8a).